etna/datasets/hierarchical_structure.py

```diff
@@ -0,0 +1,186 @@
+from collections import defaultdict
+from itertools import chain
+from queue import Queue
+from typing import Dict
+from typing import List
+from typing import Optional
+from typing import Union
+
+from scipy.sparse import csr_matrix
+from scipy.sparse import lil_matrix
+
+from etna.core import BaseMixin
+
+
+class HierarchicalStructure(BaseMixin):
+    """Represents hierarchical structure of TSDataset."""
+
+    def __init__(self, level_structure: Dict[str, List[str]], level_names: Optional[List[str]] = None):
+        """Init HierarchicalStructure.
+
+        Parameters
+        ----------
+        level_structure:
+            Adjacency list describing the structure of the hierarchy tree (i.e. {"total":["X", "Y"], "X":["a", "b"], "Y":["c", "d"]}).
+        level_names:
+            Names of levels in the hierarchy in the order from top to bottom (i.e. ["total", "category", "product"]).
+            If None is passed, level names are generated automatically with structure "level_<level_index>".
+        """
+        self.level_structure = level_structure
+        self._num_nodes = self._find_num_nodes(self.level_structure)
+        self._hierarchy_root = self._find_tree_root(self.level_structure)
+
+        hierarchy_levels = self._find_hierarchy_levels(self.level_structure)
+        tree_depth = len(hierarchy_levels)
+
+        self.level_names = self._get_level_names(level_names, tree_depth)
+        self._level_series: Dict[str, List[str]] = {self.level_names[i]: hierarchy_levels[i] for i in range(tree_depth)}
+        self._level_to_index: Dict[str, int] = {self.level_names[i]: i for i in range(tree_depth)}
+
+        self._segment_num_reachable_leafs: Dict[str, int] = self._get_num_reachable_leafs(hierarchy_levels)
+
+        self._segment_to_level: Dict[str, str] = {
+            segment: level for level in self._level_series for segment in self._level_series[level]
+        }
+
+    @staticmethod
+    def _get_level_names(level_names: Optional[List[str]], tree_depth: int) -> List[str]:
+        """Assign level names if not provided."""
+        if level_names is None:
+            level_names = [f"level_{i}" for i in range(tree_depth)]
+
+        if len(level_names) != tree_depth:
+            raise ValueError("Length of `level_names` must be equal to hierarchy tree depth!")
+
+        return level_names
+
+    @staticmethod
+    def _find_tree_root(adj_list: Dict[str, List[str]]) -> str:
+        """Find hierarchy top level (root of tree)."""
+        children = set(chain(*adj_list.values()))
+        parents = set(adj_list.keys())
+
+        tree_roots = parents.difference(children)
+        if len(tree_roots) != 1:
+            raise ValueError("Invalid tree definition: unable to find root!")
+
+        return tree_roots.pop()
+
+    @staticmethod
+    def _find_num_nodes(adj_list: Dict[str, List[str]]) -> int:
```

Comment:
Rename `adj_list` to  `hierarchy_structure` everywhere to make the naming consentient